### PR TITLE
Drop support for Python < 3.6

### PR DESCRIPTION
**Pull request by michael-k**
Most of the changes are replacing `basestring` with `str`, `assertRaisesRegexp` with `assertRaisesRegex`, and `assertEquals` with `assertEqual` and were done by 2to3.

I've excluded some of 2to3s useless changes that would just increase the PR even more. Like replacing `for pk, pv in v.items():` with `for pk, pv in list(v.items()):`.

I had to adjust `test_ref_hash` method to get the tests to pass.  I wondered why the tests on master didn't complain about it before and noticed, that only [109](https://travis-ci.org/github/cloudtools/troposphere/jobs/737421472#L980) tests are running in the Python 3 jobs on Travis (and on my machine).  Compared to [447](https://travis-ci.org/github/cloudtools/troposphere/jobs/737421468#L727) in the Python 2 job.
With this PR all tests are running with Python 3.

----

First step towards cloudtools/troposphere#1558 (troposphere 3.0)

Closes cloudtools/troposphere#1091, cloudtools/troposphere#1295, cloudtools/troposphere#1299

### CODE REVIEW

File: tests/test_basic.py [hunk 1]
```diff
@@ -381,7 +381,7 @@ def test_ref_eq(self):
 
     def test_ref_hash(self):
         s = hash("AWS::NoValue")
-        r = hash(Ref(s))
+        r = hash(Ref("AWS::NoValue"))
```
Comment (michael-k) [inline tests/test_basic.py:384]: This worked on Python 2, but failed on Python 3.

Python 2.7.18:

```python
>>> hash("AWS::NoValue")
-6577466268859890267
>>> hash(hash("AWS::NoValue"))
-6577466268859890267
```

Python 3.8.6

```python
>>> hash("AWS::NoValue")
-3204540641682287495
>>> hash(hash("AWS::NoValue"))
-898697632468593544
>>> hash(hash(hash("AWS::NoValue")))
-898697632468593544
```

ref cloudtools/troposphere#1053

**Comment by PatMyron** [comment]
> Most of the changes [...] where done by 2to3.
> I've excluded some of 2to3s useless changes

Might be nice to separate out `2to3`'s automatic changes and manual changes into separate commits

**Comment by michael-k** [comment]
> Might be nice to separate out manual changes and `2to3`'s automatic changes into separate commits

Sure, I'll try to make it easier to review this PR.  (Later today, it's already past midnight here ;) )

**Comment by michael-k** [comment]
I've the changes up in smaller commits.  The first one is still pretty massive.  It might be best to run `2to3 -n -w --no-diffs .` on the current master and check the diff between that and my first commit.  Just to make sure that I didn't smuggle in any changes.  And then review the other commits one by one.

**Comment by markpeek** [comment]
@michael-k per your comment in the awacs Python3 PR, here's my plan:
There are a few more changes I want to add to troposphere for the final Python2 release. Likely done this weekend. I'll then do a final Python 2 release of 2.7.0. After that release I was going to switch the project over to Python3 only and do a 3.0.0 major release. Per your offer, let me ping back on this PR when I'm ready for the Python3 changes and you can bring this PR up to date.

After this PR has been landed, I'll look at the black conversion and switching the repo from master to main.

**Comment by markpeek** [comment]
[Release 2.7.0](https://github.com/cloudtools/troposphere/releases/tag/2.7.0) has been released and should be the last Python 2.x release. You can now rebase these changes onto it. Thanks.

**Comment by michael-k** [comment]
@markpeek done.  The PR is ready for review :)

**Comment by markpeek** [comment]
Awesome. Thanks @michael-k!

**Pull request merged**